flink-runtime/src/main/java/org/apache/flink/runtime/operators/JoinDriver.java

```diff
@@ -128,17 +130,22 @@ public void prepare() throws Exception{
 				ConfigConstants.DEFAULT_RUNTIME_HASH_JOIN_BLOOM_FILTERS);
 
 		// create and return joining iterator according to provided local strategy.
-		if (objectReuseEnabled) {
-			switch (ls) {
-				case INNER_MERGE:
-					this.joinIterator = new ReusingMergeInnerJoinIterator<>(in1, in2, 
+		if (reset) {
+			resetForIterativeTasks(in1, in2, serializer1, serializer2, comparator1, comparator2, pairComparatorFactory);
+			reset = false;
+		}
+		if (joinIterator == null) {
```

Comment:
I can submit a PR with my updated changes. As said above this time though the drivers are ResettableDrivers, `reset()` does not have any impl and everything is done in `prepare()`.